webui/react/src/types.ts

```diff
@@ -406,18 +398,33 @@ export interface Step extends WorkloadGroup, StartEndTimes {
   training: MetricsWorkload;
 }
 
-export interface Metrics {
-  numInputs?: number;
-  validationMetrics?: Record<string, number>;
+type MetricStruct = Record<string, number>;
+export interface Metrics extends Api.V1Metrics {
```

Comment:
IMO we don't get much out of the extension right now and it only adds the risk of introducing unwanted, and unpopulated, fields to the object in the future. do we use it for something?

Reply:
In `V1Metrics`, the types for `avgMetrics` and `batchMetrics` are `any` and `Array<any>`, respectively corresponding to their proto schemas of `Struct` and `repeated Struct`.

@mpkouznetsov said that they will always be of the form `Record<string, number>` and `Array<Record<string, number>>`. This extension makes that explicit. The attributes in the extension are present in the original interface, but here we are refining them.

Maybe it is bad form though to encode assumptions not present in the schema itself though? I can instead parse them according to the expected format and leave them null otherwise (even though `validationMetrics` is a required field for `Metrics`, `Metrics` itself is only ever present as an optional field).

Reply:
^^went ahead and did this, i.e. validate, parse and narrow.

Reply:
> This extension makes that explicit. 

maybe I'm reading this incorrectly I'm not sure I see how the `extends Api.V1Metrics` at the end make it more explicit compared to having the type as you want it with Record<string, number> ... and then using the decoder like you already do.

> Maybe it is bad form though to encode assumptions not present in the schema itself though?

we already do someof this and it's a tradeoff. we eventually have to make some of these assumptions to be able to use the data. what we used to do was use iotypes package to enforce the assumptions clientside. now with the new apis we don't need as much of that but when we do we do the check in the decoder module (IIRC)

Reply:
> what we used to do was use iotypes package to enforce the assumptions clientside. now with the new apis we don't need as much of that but when we do we do the check in the decoder module (IIRC)

I saw your change wrt to this and those make sense to me👍🏼

Reply:
> how the `extends Api.V1Metrics` at the end makes it more explicit 
yeah, i don't think I articulated it very well. What I meant was that:
- having a type `OurMetrics` that extends `V1Metrics` while narrowing its attributes, AND
- asserting that data that is officially of type `V1Metrics` is actually of type `OurMetrics`
makes the assumption about `V1Metrics` explicit. anyway I guess it's a moot point now.